Add keyboard-accessible means of reordering widgets

westonruter opened the pull request:
Widgets on the admin page have this, and we need it in the customizer as well.

Currently in WordPress, the feature is provided via the "Enable Accessibility" link:

> ![01-enable-accessibility-mode](https://f.cloud.github.com/assets/134745/1289782/fa6cefc2-3026-11e3-88a8-25b43d9009fa.png)

The inline widget form is disabled and an "Edit" link appears at `/wp-admin/widgets.php?widgets-access=on`:

> ![02-accessibility-mode-on-edit-link](https://f.cloud.github.com/assets/134745/1289783/fd1a8748-3026-11e3-90f9-0b85b7aafe97.png)

Then a standalone page (e.g. at `/wp-admin/widgets.php?widgets-access=on&editwidget=categories-2&sidebar=sidebar-1&key=0`) appears which contains the widget form, followed by a sidebar selector and a position within each sidebar:

> ![03-accessible-editing](https://f.cloud.github.com/assets/134745/1289787/0459dbd0-3027-11e3-8705-fcfd301a34e4.png)

There has got to be a better way to do this. The menu editor has a better model, I think:

> ![menus__wordpress_develop__wordpress](https://f.cloud.github.com/assets/134745/1289848/fab2d43c-3027-11e3-8f79-e73c79d43e1f.png)

We could add these links to the `widget-control-actions`.

Related to #26

westonruter commented:
@shaunandrews thoughts on this?

westonruter commented:
@MichaelArestad do you have the markup for your accessible widget reordering?

westonruter commented:
Markup is located here: http://codepen.io/MichaelArestad/pen/bfaa627a1b011b80c481b9399987ae6d

![widget_accessibility_features_-_codepen](https://f.cloud.github.com/assets/134745/1756796/c239d958-6677-11e3-9b05-91c0119c55d9.png)

![widget_accessibility_features_-_codepen-2](https://f.cloud.github.com/assets/134745/1756800/d153b620-6677-11e3-8473-8e70fe445784.png)

westonruter commented:
I've filed the Trac ticket for adding the hooks to `wp_widget_control` and a ticket for fixing a keyboard accessibility problem in the customizer:
- [#26661 Add before/after hooks to override output of wp_widget_control()](http://core.trac.wordpress.org/ticket/26661)
- [#26633 Customizer form submission prevention impairs accessibility of links in customizer controls](http://core.trac.wordpress.org/ticket/26633)

westonruter commented:
Thanks to @shaunandrews for [volunteering](https://irclogs.wordpress.org/chanlog.php?channel=wordpress-ui&day=2013-12-16&sort=asc#m144551) to take a stab at integrating this:

> I can take a stab at @michaelarestad's accessibility improvements — not sure how far I'll get. Does that help?

We are currently partially blocked by the [#26661](http://core.trac.wordpress.org/ticket/26661), the `wp_widget_control_before`/`wp_widget_control_after` hooks being added.

westonruter commented:
@shaunandrews we now have the temp hooks, as of [r26865](http://core.trac.wordpress.org/changeset/26865). I think they should be permanent, however, having their `temp_` prefixes removed.

westonruter commented:
@shaunandrews here's my work-in-progress: https://gist.github.com/westonruter/8396525

westonruter commented:
@MichaelArestad I'm still working on integration of your work. I have the Reorder link now appearing, and when clicking it, the Up/Down/Move links appear with each widget. The Up/Down arrows now reorder widgets as expected. I'm still working on the move widget area.

Can you advise on a focus/hover state for `.move-widget`, `.move-widget-down`, and `.move-widget-up`?

/cc @shaunandrews

MichaelArestad commented:
Awesome! Was thinking about this last night. I would do a simple color change for just the icon making it a bit darker. Also, since this is the customizer version and isn't really needed to be mobile friendly, I would remove the dotted borders altogether allowing them to take up less space as a whole and look like they have more room. Maybe something like this: http://codepen.io/MichaelArestad/pen/JbjqA

westonruter committed:
Integrate keyboard-accessible widget reordering

Props @MichaelArestad

* Adds Reorder link next to Add Widget link
* Reorder link causes reorder buttons with each widget to appear
* Widgets can be moved up/down via buttons
* Widgets can be moved to other sidebars via new subpanel (not yet fully functional)

Other changes:
* Use Backbone for keeping track of which sidebars are rendered
* Add tabIndex to new UI elements for keyboard accessibility
* Add methods for obtaining a widget’s sidebar control,
  and a sidebar’s widget controls
* Add widget control methods for moving a widget up/down,
  and for getting its position in the sidebar.

westonruter commented:
@MichaelArestad I pushed up the initial integration; check out the `accessible-widget-reordering` branch to see it. It is functional, except for an issue with moving a widget to another sidebar. I've also added the dark color for when the links are focused/hovered. I didn't remove the borders yet.

Some todos before this PR can be merged:
- [x] the widget control does not get properly reinitialized yet when moved to another sidebar
- [x] listen for changes to which sidebars are currently rendered, and hide these sidebars from being shown among the `.widget-area-select`
- [x] when moving a widget to a new sidebar, expand the section for the new destination sidebar and re-open the widget for customizing.
- [x] remove borders from buttons in the `.widget-reorder-nav`

westonruter committed:
Prevent empty widget addition from being flagged as unsanitary

westonruter committed:
Prevent widget element lookup race condition when moving widget to another sidebar

* scope widget element lookups by associated sidebar
* instead of appending to the parent, we should append relative to the first widget found

westonruter committed:
Improve logic for cleaning up when widget is removed from sidebar

* Remove required preview refresh when widget is moved to another sidebar
* Ensure that widget moved to other sidebar without drag/drop isn’t trashed

MichaelArestad committed:
Added more icon matchers. Cleaned up CSS a bit.

Cool.

MichaelArestad committed:
Fixed a few issues with tab focusing

westonruter committed:
Merge pull request #75 from MichaelArestad/accessible-widget-reordering

Add more icon matchers, clean up CSS, improve focus states

westonruter committed:
Fix live previews for moving widget to another sidebar

* Send all sidebars_widgets settings, not just the setting for the widget being edited
* Let incoming customized POST data completely override sidebars_widgets settings; stop merging
* Stop deleting widget instance settings from preview customizer because unnecessary and
  because of possible race condition when moving widgets across sidebars

westonruter commented:
@MichaelArestad with fba1988 I've fixed the architectural problem that was preventing your keyboard-accessible widget ordering from working. Now I just heed to do two more tasks (above) and this PR should be complete, and they're just UI improvements.

westonruter committed:
Update availability of target sidebars when their render state changes

westonruter committed:
Focus on newly moved widget control

Focus on first focusable instead of on first visible input

westonruter commented:
This feature is ready for review.

westonruter committed:
Add theme support for twentyfourteen
